### PR TITLE
ENH: Cleanup npy_find_array_wrap

### PR DESCRIPTION
**Pull request by eendebakpt**
- Use `NPY_PRIORITY` instead of `0` where appropriate
- Simplify implementation by making use of `Py_NewRef`
- Eliminate `wrap == NULL` check

### CODE REVIEW

**Comment by seberg** [comment]
Is it worth it?  I had done this intentionally to not change behavior.  I don't _mind_ changing behavior here (who uses `-inf` as priority...).
IIRC, we tried relying on this in the PR, but it didn't end up nicer so didn't bother.  So if it ends up nicer or faster, I am OK with doing this (with a release note maybe).

**Comment by eendebakpt** [comment]
Not sure yet whether it is much faster, but the method showed up in profiling. I will do some more testing

**Comment by seberg** [comment]
> but the method showed up in profiling

Ah, interesting.  I would have thought it doesn't matter much (besides the attribute lookups).

**Review by seberg — COMMENTED** [review]
I am OK with this, but it is a (very) subtle change in behavior.  IIRC, @mhvk was also leaning towards this being nice just because it is a bit easier to reason about (and `-inf` is a weird array-priority).

Maybe we can actually add a release note that `-inf` array priority now causes the array-wrap to never be picked?

Otherwise, a small nit, `DBL_MAX` is not the right choice, I think. Using NewRef is nice of course :)!

**Comment by mhvk** [comment]
> Maybe we can actually add a release note that `-inf` array priority now causes the array-wrap to never be picked?

I don't think this is true even in the revised implementation: a single non-array argument will still get picked because of the `wrap == NULL` case. I think the difference is just if there are two arguments with different but negative `__ARRAY_PRIORITY__`, then the less negative one will get picked, not the first. That seems better!

Though I am actually fine with removing the last remaining `wrap == NULL` clause!

File: numpy/_core/src/multiarray/arraywrap.c [hunk 1]
```diff
@@ -34,7 +33,7 @@ npy_find_array_wrap(
     PyObject *wrap = NULL;
     PyObject *wrap_type = NULL;
 
-    double priority = -DBL_MAX;  /* silence uninitialized warning */
+    double priority = -NPY_INFINITY;  /* silence uninitialized warning */
```
Comment (mhvk) [inline numpy/_core/src/multiarray/arraywrap.c:36]: Hmm, actually the comment should go - now one really would have a problem if the variable were uninitialized!

**Comment by eendebakpt** [comment]
@mhvk The behaviour is unchanged with this PR. Also with current main for two arrays with negative priority the least negative gets selected. I added a test for this (it passed on current main already).

I left the `wrap == NULL` clause, since for performance it hardly matters and if we remove it we would have to document the fact that -INF priority does not work.  For performance reasons we should eliminate the increfs and decrefs to Py_None, but also there the gain is minimal.

**Comment by mhvk** [comment]
Ah, yes, I now see that my logic was incorrect - on current main the initial value of priority gets overwritten by the first array wrap found (when `wrap == NULL`).

**Review by mhvk — COMMENTED** [review]
Two tiny little things left. Thanks!

File: numpy/_core/src/multiarray/arraywrap.c [hunk 2]
```diff
@@ -68,10 +66,9 @@ npy_find_array_wrap(
             double curr_priority = PyArray_GetPriority(obj, 0);
             if (wrap == NULL || priority < curr_priority
```
[the remark this answers is not in the document]
Comment (mhvk) [reply numpy/_core/src/multiarray/arraywrap.c:67]: I thought that would be wanted, but perhaps not. Maybe best to just keep it as you have it!

**Review by mhvk — APPROVED** [review]
Thanks, looks all good now! @charris - I forget if numpy is OK with squash-and-merge or whether we want a rebase to get 1 commit?

**Comment by charris** [comment]
> OK with squash-and-merge

It's fine, and preferred when there are many commits. I don't use it for backports because `git branch --merged` doesn't detect it, but it is fine in main. I just wish there was some indication on the page when it is used so I would know what hash to use for cherry-picking :)

**Pull request merged**